### PR TITLE
Refactor cipher modules

### PR DESCRIPTION
**Pull request by qcam**

### CODE REVIEW

File: lib/mix/tasks/secrex.encrypt.ex [hunk 1]
```diff
@@ -14,11 +14,13 @@ defmodule Mix.Tasks.Secrex.Encrypt do
...
-      encrypted = encrypt(path, key)
+      Mix.shell().info("Encrypting #{path}")
+
+      {:ok, encrypted} = encrypt(path, key)
+
       File.write!(encrypted_path(path), encrypted)
     end
 
-    Mix.shell().info("files encrypted")
+    Mix.shell().info("Files encrypted")
```
Comment (lexmag) [inline lib/mix/tasks/secrex.encrypt.ex]:
```suggestion
    Mix.shell().info("Files have been encrypted")
```
To match the counterpart message?

File: lib/secrex/aes.ex [hunk 2]
```diff
@@ -1,18 +1,46 @@
...
+    {:ok, init_vector <> tag <> encrypted}
   end
 
-  def decrypt(encrypted, key) do
-    key = hash(key)
-    <<iv::binary-16, tag::binary-16, encrypted::binary>> = encrypted
-    :crypto.block_decrypt(:aes_gcm, key, iv, {"AES256GCM", encrypted, tag})
+  @impl true
+  def decrypt(ciphertext, key) do
+    hashed_key = hash(key)
```
Comment (lexmag) [inline lib/secrex/aes.ex]:
```suggestion
    key_digest = hash(key)
```

File: lib/secrex/aes.ex [hunk 3]
```diff
@@ -1,18 +1,46 @@
...
+  # Additional Authenticated Data.
+  @aad "AES256GCM"
+
+  @iv_length 16
+  @tag_length 16
+
+  @impl true
+  def encrypt(plaintext, key) do
+    init_vector = initialize_vector()
+    hashed_key = hash(key)
```
Comment (lexmag) [inline lib/secrex/aes.ex]:
```suggestion
    key_digest = hash(key)
```

File: lib/secrex/aes.ex [hunk 4]
```diff
@@ -1,18 +1,46 @@
...
+
+  # Additional Authenticated Data.
+  @aad "AES256GCM"
+
+  @iv_length 16
+  @tag_length 16
+
+  @impl true
+  def encrypt(plaintext, key) do
+    init_vector = initialize_vector()
```
Comment (lexmag) [inline lib/secrex/aes.ex]:
```suggestion
    init_vector = initialize_vector(@iv_length)
```

File: lib/secrex/aes.ex [hunk 5]
```diff
@@ -1,18 +1,46 @@
...
-  def decrypt(encrypted, key) do
-    key = hash(key)
-    <<iv::binary-16, tag::binary-16, encrypted::binary>> = encrypted
-    :crypto.block_decrypt(:aes_gcm, key, iv, {"AES256GCM", encrypted, tag})
+  @impl true
+  def decrypt(ciphertext, key) do
+    hashed_key = hash(key)
+
+    case ciphertext do
+      <<init_vector::binary-size(@iv_length), tag::binary-size(@tag_length), encrypted::binary>> ->
```
Comment (lexmag) [inline lib/secrex/aes.ex]:
```suggestion
      <<init_vector::size(@iv_length)-bytes, tag::size(@tag_length)-bytes, encrypted::binary>> ->
```

File: lib/mix/tasks/secrex.decrypt.ex [hunk 6]
```diff
@@ -15,11 +15,16 @@ defmodule Mix.Tasks.Secrex.Decrypt do
...
+      Mix.shell().info("Decrypting #{enc_path}")
+
+      case decrypt(enc_path, key) do
+        {:ok, decrypted} ->
+          File.write!(path, decrypted)
+          Mix.shell().info("Files have been decrypted")
 
-    Mix.shell().info("files decrypted")
+        {:error, reason} ->
+          Mix.raise("Could not decrypt file, reason: " <> inspect(reason))
```
Comment (lexmag) [inline lib/mix/tasks/secrex.decrypt.ex]:
```suggestion
          Mix.raise("Cannot decrypt file, reason: " <> inspect(reason))
```

File: lib/mix/tasks/secrex.decrypt.ex [hunk 7]
```diff
@@ -15,11 +15,16 @@ defmodule Mix.Tasks.Secrex.Decrypt do
...
-    end
+      Mix.shell().info("Decrypting #{enc_path}")
+
+      case decrypt(enc_path, key) do
+        {:ok, decrypted} ->
+          File.write!(path, decrypted)
+          Mix.shell().info("Files have been decrypted")
 
-    Mix.shell().info("files decrypted")
+        {:error, reason} ->
```
Comment (lexmag) [inline lib/mix/tasks/secrex.decrypt.ex:25]: We either should make `reason` be atom or not inspect it when raising.

**Review by lexmag — DISMISSED** [review]

File: lib/mix/tasks/secrex.decrypt.ex (hunk 7 above)
Comment (qcam) [reply lib/mix/tasks/secrex.decrypt.ex:25]: converted reason to atom.

**Review by lexmag — APPROVED** [review]

**Pull request merged**